File: go/1639. Number of Ways to Form a Target String Given a Dictionary.go
func numWays(words []string, target string) int {
	wordLength := len(words[0])
	if len(target) > wordLength {
		return 0
	}
	db := make([][]int, wordLength)
	for i := range db {
		db[i] = make([]int, 26)
	}
	for i := 0; i < wordLength; i++ {
		for _, word := range words {
			index := int(word[i] - 'a')
			db[i][index]++
		}
	}
	cache := make([][]int, len(target))
	for i := range cache {
		cache[i] = make([]int, wordLength)
		for j := range cache[i] {
			cache[i][j] = -1
		}
	}
	var find func(int, int, [][]int) int
	find = func(targetIndex, index int, cache [][]int) int {
		if targetIndex >= len(target) {
			return 1
		}
		if index >= wordLength {
			return 0
		}
		if cache[targetIndex][index] != -1 {
			return cache[targetIndex][index]
		}
		answer := find(targetIndex, index+1, cache)
		answer += db[index][int(target[targetIndex]-'a')] * find(targetIndex+1, index+1, cache)
		answer %= 1000000007
		cache[targetIndex][index] = answer
		return answer
	}
	return find(0, 0, cache)
}

File: java/1639. Number of Ways to Form a Target String Given a Dictionary.java
class Solution {
    private static final int MOD = (int)(1e9) + 7;
    public int numWays(String[] words, String target) {
        int wordLength = words[0].length();
        if (wordLength < target.length()) return 0;
        int[][] wordDb = new int[wordLength][26];
        for (String w : words) {
            for (int i = 0; i < wordLength; i++) {
                wordDb[i][w.charAt(i)-'a']++;
            }
        }
        Integer[][] cache = new Integer[wordLength][target.length()];
        return helper(target, wordDb, 0, 0, cache);
    }
    private int helper(
        String target,
        int[][] wordDb,
        int targetIndex,
        int dbIndex,
        Integer[][] cache
    ) {
        if (targetIndex == target.length()) return 1;
        if (dbIndex == wordDb.length) return 0;
        if (cache[dbIndex][targetIndex] != null) return cache[dbIndex][targetIndex];
        long answer = helper(target, wordDb, targetIndex, dbIndex+1, cache);
        long ways = wordDb[dbIndex][target.charAt(targetIndex)-'a'];
        answer += ways * helper(target, wordDb, targetIndex+1, dbIndex+1, cache);
        answer %= MOD;
        return cache[dbIndex][targetIndex] = (int)answer;
    }
}
